### PR TITLE
SAK-49217 - Accept Link: headers with no space after the comma

### PR DESCRIPTION
**Pull request by csev**

### CODE REVIEW

**Review by ottenhoff — APPROVED** [review]

**Review by jonespm — APPROVED** [review]
I didn't test this against Canvas but this looks good and passes the tests. I wonder what Canvas would have changed with this as it wasn't something we noticed. in Python the `requests `package provides these links directly without any coding required and hasn't had an issue. I've never seen anything quite the same in Java.

File: basiclti/tsugi-util/src/test/org/tsugi/http/HttpUtilTest.java [hunk 1]
```diff
@@ -45,5 +45,13 @@ public void testLinkExtractRel() {
 		String differencesUrl = HttpUtil.extractLinkByRel(headerValues, "differences");
 		assertEquals("https://lms.example.com/sections/2923/memberships?since=1422554502", differencesUrl);
 
+		// The Canvas way of Multiple HTTP headers with out a space after the comma
+		// This confuses the java.util.net.HttpHeaders.allValues()
+		// So we compensate.
+		headerValues = new ArrayList<String>();
+		headerValues.add("<https://umich.instructure.com/api/lti/courses/624214/names_and_roles?page=1&per_page=50>; rel=\"current\",<https://umich.instructure.com/api/lti/courses/624214/names_and_roles?page=2&per_page=50>; rel=\"next\",<https://umich.instructure.com/api/lti/courses/624214/names_and_roles?page=1&per_page=50>; rel=\"first\",<https://umich.instructure.com/api/lti/courses/624214/names_and_roles?page=3&per_page=50>; rel=\"last\"");
```
Comment (jonespm) [inline basiclti/tsugi-util/src/test/org/tsugi/http/HttpUtilTest.java:52]: I see you're including this per_page in the test cases. Is this configurable in the code? I think it having it configurable would make it easier to test (setting this value lower) and it should be set to the max by default for performance. [The documentation says there is an unspecified limit](https://canvas.instructure.com/doc/api/file.pagination.html) on how how big this value can be set but generally it seems like most API's it's 200.

Comment (ern) [reply basiclti/tsugi-util/src/test/org/tsugi/http/HttpUtilTest.java:52]: +1

**Comment by jonespm** [comment]
Followup from the call this might be possible to use this [org.apache.http.message.BasicHeaderValueParser ](https://hc.apache.org/httpcomponents-core-4.4.x/current/httpcore/apidocs/org/apache/http/message/BasicHeaderValueParser.html) library. I also believe that switching to org.apache.http.client.HttpClient instead of java.net.http.HttpClient and using their `response.getHeaders("Link")` might also be using this same library and produce an expected response.

File: basiclti/tsugi-util/src/test/org/tsugi/http/HttpUtilTest.java (hunk 1 above)
Comment (csev) [reply basiclti/tsugi-util/src/test/org/tsugi/http/HttpUtilTest.java:52]: These test cases are just using the exact strings including key value parameters.  These parameters are not interpreted at all in the unit test.   And yes - I have seen defaults of 200 and defaults of 50.  The key when building SakaiPlus is not to assume "infinity" - which the Sakai code does handle if the Link: headers are formatted in a way that can be parsed.

**Comment by csev** [comment]
So @jonespm - I tried both the 4.x and 5.x `BasicHeaderValueParser` and they are both pretty lame.  Their use case is for headers that look like this:

`param1=value1; param2=value2; param3=value3`

When presented any kind of link header with pretty intricate rules about separators, url encoding, etc, both of these libraries fall flat and get 100% lost and can't even do the basic stuff that Java does until it meets the Canvas variant of the header.

I dove into the code thinking that to avoid the dreaded regex, I would do some ad hoc parsing.   I started to rough it out and figured that a simple regex is a clearer and safer path.  I suggest merging as it.

param1=value1; param2=value2; param3=value3

**Comment by ottenhoff** [comment]
> I suggest merging as it.

+1

**Comment by csev** [comment]
Thanks.

**Pull request merged**